### PR TITLE
Add Keycloak support to docker-compose sub-generator

### PR DESCRIPTION
**Pull request by mraible**
- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary. See https://github.com/jhipster/jhipster.github.io/pull/493.
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
I figured out how to get the files generated. Now I'm experiencing an issue where login to Keycloak doesn't work when running in a Docker container. It works fine if the app runs outside of Docker. The error can be viewed in the screenshot below.

screen shot 2017-10-01 at 8 38 57 pm


I found a [similar problem](https://stackoverflow.com/questions/31383514/unable-to-get-enableoauth2sso-working-badcredentialsexception-could-not-obta) on Stack Overflow and tried changing JSESSIONID to APPSESSIONID (as suggested [here](https://stackoverflow.com/a/42303754/65681)), but it doesn't work.

**Comment by jdubois** [comment]
@mraible I'm running your PR right now:

- I don't have your error page
- I am forwarded to Keycloak and I still have the `Invalid parameter: redirect_uri` message

**Comment by mraible** [comment]
@jdubois Can you look at the URL and see what redirect_uri is being sent?

**Comment by jdubois** [comment]
@mraible yes, here is the full URL:

    http://localhost:9080/auth/realms/jhipster/protocol/openid-connect/auth?client_id=web_app&redirect_uri=http://127.0.0.1:8080/login&response_type=code&scope=openid%20profile%20email&state=AOKEkZ

**Comment by mraible** [comment]
The problem is `redirect_uri=http://127.0.0.1:8080/login` - that should be `redirect_uri=http://localhost:8080/login`. Or we should modify Keycloak to allow this as a redirect_uri.

**Comment by danielpetisme** [comment]
I made an "accept everything" configuration. It's really dangerous and dedicated for dev environments.
capture d ecran 2017-10-03 a 01 29 26
It works with any host name (localhost, loopback IP or LAN IP or other) and ports (the actual committed version only work with port `8080`).
We must link the official doc (or provide a gist) for a real-life config.
@mraible @jdubois If your test run with this config I can provide the json config file for keycloak via PR.

**Comment by mraible** [comment]
@danielpetisme I tried your settings and I still get the same issue. The strange thing is I can start PostgreSQL and Keycloak using `docker-compose`, and then my app with `mvn -Pprod`, and everything works fine.

**Comment by mraible** [comment]
I turned on debug logging for Spring Security. Here's what the error log shows when trying to login and running everything using `docker-compose -f src/main/docker/app.yml up`:

```
blog-app_1         | Caused by: org.springframework.web.client.ResourceAccessException: I/O error on POST request for "http://localhost:9080/auth/realms/jhipster/protocol/openid-connect/token": Connection refused (Connection refused); nested exception is java.net.ConnectException: Connection refused (Connection refused)
blog-app_1         | 	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:666)
blog-app_1         | 	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:621)
blog-app_1         | 	at org.springframework.security.oauth2.client.token.OAuth2AccessTokenSupport.retrieveToken(OAuth2AccessTokenSupport.java:137)
blog-app_1         | 	... 75 common frames omitted
blog-app_1         | Caused by: java.net.ConnectException: Connection refused (Connection refused)
blog-app_1         | 	at java.net.PlainSocketImpl.socketConnect(Native Method)
```

Full log at https://gist.github.com/mraible/1eab6f5c82b04efa8495f126fbf5a138.

**Comment by sdoxsee** [comment]
Hi @mraible I think you need to refer to the service name "keycloak" instead of localhost when running your spring boot app in docker-compose...
http://keycloak:9080/auth/realms/jhipster/protocol/openid-connect/token

**Comment by mraible** [comment]
Yeah, I tried that. It doesn't work. See same suggestion and my results on JHipster-dev mailing list:

https://groups.google.com/forum/m/#!topic/jhipster-dev/VwiULdJGuTM

**Comment by sdoxsee** [comment]
Add under blog_app service

links:
            - keycloak

**Comment by jdubois** [comment]
I can't test right now, but it's either:

- you're "inside" Docker and the hostname is the name of your Docker image: probably "http://keycloak:..."
- you are outside (like a user accessing the app) and then it's "localhost" and you need to expose the port

File: generators/docker-compose/templates/realm-config/_jhipster-realm.json [hunk 1]
```diff
@@ -0,0 +1,1705 @@
...
+        "jsonType.label" : "String"
+      }
+    } ],
+    "useTemplateConfig" : false,
+    "useTemplateScope" : false,
+    "useTemplateMappers" : false
+  }, {
+    "id" : "9080f291-17ef-4022-8afe-7822595006c6",
+    "clientId" : "web_app",
+    "rootUrl" : "http://localhost:8080",
```
Comment (pascalgrimaud) [inline generators/docker-compose/templates/realm-config/_jhipster-realm.json]: I don't think `localhost` can work here

File: generators/docker-compose/templates/realm-config/_jhipster-realm.json [hunk 2]
```diff
@@ -0,0 +1,1705 @@
...
+    "surrogateAuthRequired" : false,
+    "enabled" : true,
+    "clientAuthenticatorType" : "client-secret",
+    "secret" : "web_app",
+    "redirectUris" : [ "http://localhost:8080/*" ],
```
Comment (pascalgrimaud) [inline generators/docker-compose/templates/realm-config/_jhipster-realm.json]: same here

File: generators/docker-compose/templates/realm-config/_jhipster-realm.json [hunk 3]
```diff
@@ -0,0 +1,1705 @@
...
+    "webOrigins" : [ "http://localhost:8080/*" ],
```
Comment (pascalgrimaud) [inline generators/docker-compose/templates/realm-config/_jhipster-realm.json]: same here

**Comment by pascalgrimaud** [comment]
@mraible : can you try these configs ? It works for me.
But I had to add keycloak to my `/etc/hosts`

**docker-compose.yml** :
```yaml
version: '2'
services:
    ojo-app:
        image: ojo
        environment:
            - SPRING_PROFILES_ACTIVE=prod,swagger
            - SPRING_DATASOURCE_URL=jdbc:mysql://ojo-mysql:3306/ojo?useUnicode=true&characterEncoding=utf8&useSSL=false
            - JHIPSTER_SLEEP=10
            - OAUTH2_ISSUER=http://keycloak:9080/auth/realms/jhipster
        ports:
            - 8080:8080
    ojo-mysql:
        image: mysql:5.7.19
        environment:
        - MYSQL_USER=root
        - MYSQL_ALLOW_EMPTY_PASSWORD=yes
        - MYSQL_DATABASE=ojo
        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp
    keycloak:
        extends:
            file: keycloak.yml
            service: keycloak
```

**keycloak.yml** :
```yaml
version: '2'
services:
    keycloak:
        image: jboss/keycloak:3.3.0.CR1
        command: ["-b", "0.0.0.0", "-Dkeycloak.migration.action=import", "-Dkeycloak.migration.provider=dir", "-Dkeycloak.migration.dir=/opt/jboss/keycloak/realm-config", "-Dkeycloak.migration.strategy=OVERWRITE_EXISTING", "-Djboss.socket.binding.port-offset=1000"]
        volumes:
        - ./realm-config:/opt/jboss/keycloak/realm-config
        environment:
        - KEYCLOAK_USER=admin
        - KEYCLOAK_PASSWORD=admin
        ports:
        - 9080:9080
        - 9443:9443
        - 10990:10990
```

**Comment by sdoxsee** [comment]
wouldn't adding depends_on: - keycloak under ojo-app eliminate the need for
/etc/hosts changes?

**Comment by pascalgrimaud** [comment]
I tried the `depends_on` but had the same initial issue.
If you have a config with `depends_on` which works, don't hesitate to share :)

**Comment by mraible** [comment]
Thanks @pascalgrimaud - those settings work for me too! It stinks that you have to modify your hosts file to add the following.

```
127.0.0.1	keycloak
```

**Comment by mraible** [comment]
Added documentation at https://github.com/jhipster/jhipster.github.io/pull/493

**Comment by jdubois** [comment]
Yes that host file is really annoying!!!!! Anyway, let's go for a first release, and improve after that.

**Comment by jdubois** [comment]
@mraible you have changes from other PRs, can you rebase your work?

**Comment by mraible** [comment]
@jdubois It all looks good to me -> "This branch has no conflicts with the base branch"

**Pull request merged**